### DISCUSSION TITLE
Can I make parts of my private repository public?

### DISCUSSION THREAD

**Discussion by skeddles** [community/community#22382] [New to GitHub]
<p>I have my website as a private repository. I want to make parts of it public, but still keep most of it private. It’s a huge pain to keep the inner ones up to date with my main repository.</p>
<p>Currently i just have multiple repositories nested under my main one, listed separately on github. Is there a way I could choose certain files / folders to make public / private, or maybe give certain users permission to view those folders?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>No, you cannot have different public/private settings within one repository.</p>
<p>The closest thing I can think of would be to keep the repositories separate and use submodules to create a unified directory tree, but I can’t judge if that makes sense for your workflow.</p>

**Comment by Shadrock** [comment]
<p>I would love to see this elevated to a request: my use case is that, as a professor delivering online content to students and teachers, folder-level privacy would greatly facilitate my ability to put an entire course together (e.g. labs along with lab solutions) and make it available to students, without also giving away the answers to labs. Is there an appropriate place to make this a feature request?</p>

**Comment by airtower-luna** [comment]
<p>I’m afraid that is fundamentally incompatible with the way git works. If someone can clone your repository, they get all data within it.</p>
<p>So even if Github was to implement such a thing in the web interface your students would just need to clone the repository to get the answers. Which is why I doubt Github would even try. :wink:</p>
<p>A possible solution would be to keep your answers in a private repository, and (automatically) create a public one from that which only contains the question part.</p>

**Reply by drmcnelson** [reply]
No that does not have to be fundamental to the way github works.   It is a matter of choosing to provide the feature or not.   A simple implementation for example, would be a file .gitprivate to exclude the private directory  or files from public access.

**Reply by vighnesh153** [reply]
@drmcnelson GitHub is built on top of Git. When someone clones a repository using the `git clone` command, I believe there is absolutely no interception by GitHub in that process. Therefore, this feature request should ideally be for Git, and once it's built into Git, it should be straightforward to extend to GitHub's web interface.

**Comment by Shadrock** [comment]
<p>Thanks for the suggestion. I realize this wasn’t really in keeping with a true “Git” workflow, but it so many people use Github in different ways I thought it was worth asking. I didn’t thinking about creating a duplicate without the solutions: total face-palm on my part. That’s a great work-around so thanks for bringing it up!</p>

**Comment by pahaz** [comment]
<p>You can use a trick. You can automatically create a public part of your private repository.</p>
<p>For example: you can use <a href="https://github.com/open-condo-software/gitexporter" rel="noopener nofollow ugc">gitexporter</a> command. <code>gitexporter</code> can create a new git repo with only allowed files and the same commit history.</p>
<ol>
<li>Create <code>config.json</code>:</li>
</ol>
<pre><code class="lang-auto">{
    "forceReCreateRepo": true,
    "targetRepoPath": "my-open-source-repo",
    "sourceRepoPath": ".",
    "allowedPaths": ["*"],
    "ignoredPaths": ["secrets/*"]
}
</code></pre>
<p>Run command <code>npx gitexporter config.json</code>.</p>
<p>Folder <code>my-open-source-repo</code> is a new git repo that does not include <code>"secrets/*"</code> files</p>
<p>Then you can push <code>my-open-source-repo</code> to a new open-source GitHub repository.</p>
<p>Like so:</p>
<pre><code class="lang-auto">cd my-open-source-repo
git remote set-url origin new.github.com/company/open-repo-name
git push origin master
</code></pre>
